Rules/UseBOMForUnicodeEncodedFile.cs

```diff
@@ -0,0 +1,146 @@
+﻿//
+// Copyright (c) Microsoft Corporation.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+using System.ComponentModel.Composition;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Linq;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// AvoidAlias: Check if help file uses utf8 encoding
+    /// </summary>
+    [Export(typeof(IScriptRule))]
+    public class UseBOMForUnicodeEncodedFile : IScriptRule
+    {
+        /// <summary>
+        /// AnalyzeScript: For a file that has BOM missing, check if content is encoded in ASCII
+        /// </summary>
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            byte[] byteStream = File.ReadAllBytes(fileName);
+
+            if (null == GetByteStreamEncoding(byteStream))
+            {
+                // Did not detect the presence of BOM
+                // Make sure there is no byte > 127 (0x7F) to ensure file is ASCII encoded
+                // Else emit rule violation
+
+                var byteList = byteStream.Where(o => o > 0x7F).ToList();
+                if (0 != byteList.Count)
+                { 
+                    yield return new DiagnosticRecord(string.Format(CultureInfo.CurrentCulture, Strings.UseBOMForUnicodeEncodedFileError, System.IO.Path.GetFileName(fileName), null),
+                                null, GetName(), DiagnosticSeverity.Warning, fileName);
+                }
+            }
+        }
+
+        /// <summary>
+        /// GetByteStreamEncoding: Detect the file encoding using the file's byte stream
+        /// </summary>
+        private Encoding GetByteStreamEncoding(byte[] byteStream)
+        {
+            //// Analyze BOM
```

Comment:
I think it may be cleaner to use Encoding.Unicode.GetPreamble() method: https://msdn.microsoft.com/en-us/library/system.text.unicodeencoding.getpreamble(v=vs.90).aspx